### ISSUE TITLE
`sortByRank` can be given a compare function causing inconsistent placement of `undefined`s

### ISSUE THREAD

**Issue by erights** [Agoric/Agoric#4260]
POTENTIAL SECURITY BUG

`sortByRank` currently uses `Array.prototype.sort` directly, and so only works correctly when given a `compare` function that considers `undefined` strictly bigger (`>`) than everything else. This is because `Array.prototype.sort` bizarrely moves all `undefined`s to the end of the array regardless, without consulting the `compare` function. This is a genuine bug for us NOW because sometimes we sort in reverse order by passing a reversed rank comparison function.

See https://github.com/Agoric/agoric-sdk/pull/4252 where @FUDCo fixed the forward rank comparison so that `sortByRank` applied to it would be consistent. The remaining live problem is only where we use the reversed comparison.

**Comment by mhofman** [comment]
> This is because `Array.prototype.sort` bizarrely moves all `undefined`s to the end of the array regardless, without consulting the `compare` function.

Ugh? Is this actually specified this way?

**Comment by mhofman** [comment]
Ok just read about this. Without knowing how the `sortByRank` is implemented or used, would it make sense to wrap the values of the array, and wrap any user provided compare function?

**Comment by erights** [comment]
Yes, that would be a valid fix. I just have not done any such fix yet.

**Comment by erights** [comment]
We're only concerned with known orders, where the normal forward ones (compareRank and its full ordered variants) put all undefineds at the end, and the reverse of this (compareAntiRank) that put all undefineds at the beginning. Given that, rather than do some kind of wrapper thing, simpler would be to first gather all the undefineds at the appropriate end and then sort the remainder in place. Unfortunately, `Array.prototype.sort` cannot be told to operate only on a slice of an array.

**Comment by erights** [comment]
Since we cannot ask Array.prototype.sort to operate on a slice, the next best thing is to just sort, allowing all undefineds to gather at the end. Then, if we intend to do a reverse sort, just splice the undefineds from the end to the beginning.

**Comment by aj-agoric** [comment]
considering priorities and recent activity, I am motioning to move this from "ready" to "backlog" unless someone shouts

**Comment by erights** [comment]
Attn @gibson042 @mhofman @kriskowal

**Comment by erights** [comment]
We do use an inverse of rankCompare which could have this problem. If so, we do need to fix this. Changing to ready until someone checks if this inverse rankCompare already protects itself from this problem.

Should also be moved to endo. I have no idea why I originally filed this against agoric-sdk.

**Comment by aj-agoric** [comment]
Noted, thank you @erights. Considering your comments I suggest we place a priority of P2 for now. Also I will assign this to U21 for now to ensure its considered in the next work streams(can be removed after triage).
